bookwyrm/models/bookwyrm_export_job.py

```diff
@@ -1,213 +1,445 @@
+    """a boto session that always uses settings.AWS_S3_ENDPOINT_URL"""
+
+    def client(self, *args, **kwargs):  # pylint: disable=arguments-differ
+        kwargs["endpoint_url"] = settings.AWS_S3_ENDPOINT_URL
+        return super().client("s3", *args, **kwargs)
+
+
 class BookwyrmExportJob(ParentJob):
     """entry for a specific request to export a bookwyrm user"""
 
-    export_data = FileField(null=True)
+    # Only one of these fields is used, dependent on the configuration.
+    export_data_file = FileField(null=True, storage=storage_backends.ExportsFileStorage)
+    export_data_s3 = FileField(null=True, storage=storage_backends.ExportsS3Storage)
+
+    export_json = JSONField(null=True, encoder=DjangoJSONEncoder)
+    json_completed = BooleanField(default=False)
+
+    @property
+    def export_data(self):
+        """returns the file field of the configured storage backend"""
+        # TODO: We could check whether a field for a different backend is
+        # filled, to support migrating to a different backend.
+        if settings.USE_S3:
+            return self.export_data_s3
+        return self.export_data_file
+
+    @export_data.setter
+    def export_data(self, value):
+        """sets the file field of the configured storage backend"""
+        if settings.USE_S3:
+            self.export_data_s3 = value
+        else:
+            self.export_data_file = value
 
     def start_job(self):
         """Start the job"""
-        start_export_task.delay(job_id=self.id, no_children=True)
 
-        return self
+        task = start_export_task.delay(job_id=self.id, no_children=False)
+        self.task_id = task.id
+        self.save(update_fields=["task_id"])
+
+    def notify_child_job_complete(self):
+        """let the job know when the items get work done"""
+
+        if self.complete:
+            return
+
+        self.updated_date = timezone.now()
+        self.save(update_fields=["updated_date"])
+
+        if not self.complete and self.has_completed:
+            if not self.json_completed:
+                try:
+                    self.json_completed = True
+                    self.save(update_fields=["json_completed"])
+
+                    tar_job = AddFileToTar.objects.create(
+                        parent_job=self, parent_export_job=self
+                    )
+                    tar_job.start_job()
+
+                except Exception as err:  # pylint: disable=broad-except
+                    logger.exception("job %s failed with error: %s", self.id, err)
+                    tar_job.set_status("failed")
+                    self.stop_job(reason="failed")
+
+            else:
+                self.complete_job()
+
+
+def url2relativepath(url: str) -> str:
+    """turn an absolute URL into a relative filesystem path"""
+    parsed = urlparse(url)
+    return unquote(parsed.path[1:])
+
+
+class AddBookToUserExportJob(ChildJob):
+    """append book metadata for each book in an export"""
```

Comment:
An export now consists of many separate child jobs. Is there a mechanism to prevent data races between different child jobs? Otherwise, the following can happen:

1. Start task for book B, read current JSON containing book A
2. Start task for book C, read current JSON containing book A
3. Task for B adds itself and saves; archive contains A + B
4. Task for C adds itself and saves; archive contains A + C

This means the data about book B gets lost

Reply:
hmm, I see what you mean. I guess we need something in `TriggerBookJobs` to wait for each `AddBookToUserExportJob` to finish before triggering the next one?